### PR TITLE
Remove ModelSummary validation from train loop on_trainer_init

### PR DESCRIPTION
**Pull request by ananthsub**
## What does this PR do?
- Start adding typehints to `training_loop.py` beginning with `on_trainer_init` for better static analysis w/ mypy or Pyre
- Remove duplication for `weights_summary` across `TrainLoop` and `Trainer` - the validation and property setting was happening in both places

Fixes #<issue_number>

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/trainer/trainer.py [hunk 1]
```diff
@@ -544,10 +547,7 @@ def _pre_training_routine(self):
 
         # print model summary
         if self.is_global_zero and self.weights_summary is not None and not self.testing:
-            if self.weights_summary in ModelSummary.MODES:
-                ref_model.summarize(mode=self.weights_summary)
-            else:
-                raise MisconfigurationException("weights_summary can be None, " + ", ".join(ModelSummary.MODES))
+            ref_model.summarize(mode=self.weights_summary)
```
Comment (ananthsub) [inline pytorch_lightning/trainer/trainer.py:550]: @edenafek  @awaelchli  @justusschock the model summary could be another (small) component we could split out from the core lightning module and offer as a standalone utility. the constructor for the summary object could take:
- an nn.Module
- precision
- device
- mode

connecting it back into the trainer here would be easy. it's almost there now with the `ModelSummary` object

File: pytorch_lightning/trainer/trainer.py [hunk 2]
```diff
@@ -320,6 +320,10 @@ def __init__(
         self.predict_loop = PredictLoop(self)
 
         # training state
+        if weights_summary is not None and weights_summary not in ModelSummary.MODES:
+            raise MisconfigurationException(
+                f"`weights_summary` can be None, {', '.join(ModelSummary.MODES)}, but got {weights_summary}"
+            )
```
Comment (ananthsub) [inline pytorch_lightning/trainer/trainer.py:326]: i understand we want to fail early but i dislike that the validation here is repeated again in the implementation of `LightningModule.summarize()`

File: pytorch_lightning/trainer/trainer.py (hunk 1 above)
Comment (carmocca) [reply pytorch_lightning/trainer/trainer.py:550]: Related idea: https://github.com/PyTorchLightning/pytorch-lightning/issues/4541

**Comment by carmocca** [comment]
Is this ready to review? If yes: please fill out the checklist in the original post, otherwise mark it as `[WIP]`/convert to draft

**Comment by ananthsub** [comment]
> Is this ready to review? If yes: please fill out the checklist in the original post, otherwise mark it as `[WIP]`/convert to draft

yes this is ready for review

File: pytorch_lightning/trainer/trainer.py (hunk 2 above)
Comment (awaelchli) [reply pytorch_lightning/trainer/trainer.py:326]: >  i dislike that the validation here is repeated again in the implementation of LightningModule.summarize()

It is there because this method is standalone, it has no relation to the trainer and can be used directly on the `LightningModule` outside of a training script

File: pytorch_lightning/trainer/trainer.py (hunk 1 above)
Comment (awaelchli) [reply pytorch_lightning/trainer/trainer.py:550]: Btw it's already a standalone utility. It can be used like this:

```python
summary = LayerSummary(model, mode=...)
print(summary)
```
As you say it could also support nn.Module.

**Review by awaelchli — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
LGTM !

File: pytorch_lightning/trainer/trainer.py (hunk 2 above)
Comment (ananthsub) [reply pytorch_lightning/trainer/trainer.py:326]: @awaelchli in this case shall I move this validation check for `weights_summary` to the `ModelSummary` constructor? https://github.com/PyTorchLightning/pytorch-lightning/blob/634d83134fea4bb701c24abd5a4a38adb0eddbcd/pytorch_lightning/core/memory.py#L185-L192 ? 

There is one test that asserts that the an invalid `weights_summary` will raise an exception on the Trainer init, which is why I moved this here. 

I can split this out to a separate PR if you'd like

File: pytorch_lightning/trainer/trainer.py (hunk 1 above)
Comment (ananthsub) [reply pytorch_lightning/trainer/trainer.py:550]: by standalone i mean no dependence on the lightning module. right now the precision and example inputs are looked up via attributes of `model` inside of model summary, but these could be passed in explicitly to the model summary constructor. it's not a huge deal, but something small for people to see how lightning builds reusable components

File: pytorch_lightning/trainer/trainer.py (hunk 2 above)
Comment (awaelchli) [reply pytorch_lightning/trainer/trainer.py:326]: What happens to the message in the misconfiguration error?
If using the Trainer, the message should inform about weights_summary, and when using the summary object directly it should inform about the mode.

Comment (ananthsub) [reply pytorch_lightning/trainer/trainer.py:326]: ah good catch! it makes sense to have both checks for debuggability

**Review by justusschock — APPROVED** [review]
